torch/_dynamo/variables/builder.py

```diff
@@ -2227,38 +2228,53 @@ def _automatic_dynamic(
         )
 
     # Prep for automatic dynamic
-    frame_state_entry = None
-    if name not in tx.output.frame_state:
-        # If there is no entry for this source, add the tensor to frame state with its current static size.
-        # E.g., {} -> {"x": [2, 4]}
-        frame_state_entry = FrameStateSizeEntry(None, None)
-        frame_state_entry.size = list(e.size())
+    def update_frame_state(size):
+        frame_state_entry = None
+        if name not in tx.output.frame_state:
+            # If there is no entry for this source, add the tensor to frame state with its current static size.
+            # E.g., {} -> {"x": [2, 4]}
+            frame_state_entry = FrameStateSizeEntry(None, None)
+            frame_state_entry.size = list(size)
+        else:
+            frame_state_entry = tx.output.frame_state[name]
+            if frame_state_entry.size is not None:
+                if len(size) != len(frame_state_entry.size):
+                    # If there is already an entry, and the dim mismatches, replace the frame state entry with None.
+                    # E.g. {"x": [2, 3, 4]} -> {"x": None}
+                    log.debug(
+                        "automatic dynamic %s dim %s != %s",
+                        name,
+                        len(size),
+                        frame_state_entry.size,
+                    )
+                    frame_state_entry.size = None
+                else:
+                    # If there is already an entry, and the dim matches, for every size in the frame state which
+                    # disagrees with the current static size, replace it with None. E.g., {"x": [2, 3]} -> {"x": [2, None]}
+                    for i, dim in enumerate(frame_state_entry.size):
+                        if dim is not None and size[i] != dim:
+                            log.debug(
+                                "automatic dynamic %s size(%s) %s != %s",
+                                name,
+                                i,
+                                size[i],
+                                dim,
+                            )
+                            frame_state_entry.size[i] = None
+        tx.output.frame_state[name] = frame_state_entry
+
+    if config.enable_sync_dist_compilation and dist.is_available() and dist.is_initialized():
```

Comment:
does this block get called one time per tensor argument for a compile function? it would be a lot better for collective if you can collect up all the tensor sizes into one py list first and do a single object collective over this list per rank rather than do it one time per tensor per rank.

Reply:
Yup, this is the inefficient version